Allow explicit this parameter without renaming (#4)
gilbert/es-explicit-this: README.md

> `this` isn't really a keyword, it is the natural "main" function argument in JavaScript.
> - (@spion)[https://github.com/mindeavor/es-pipeline-operator/issues/2#issuecomment-162348536]

This proposal extends the `function` declaration syntax to allow for explicit naming of what is normally called `this`.

```js
Object.defineProperty(User.prototype, {
  get: function fullName() { // original version
    return `${this.firstName} ${this.lastName}`
  },
  configurable: true,
})

// versions use the feature of this proposal

function fullName(this) { // explicit this parameter
  return `${this.firstName} ${this.lastName}`
}

function fullName(this user) { // explicit naming `this` to `user`
  return `${user.firstName} ${user.lastName}`
}

function fullName(this {firstName, lastName}) { // destructuring
  return `${firstName} ${lastName}`
}
```

> Its primary use case is to play nicely with the [function bind proposal](https://github.com/zenparsing/es-function-bind), making such functions more readable. For example:

Reviewer: ```suggestion
The original motivation of this proposal was to play nicely with the [function bind proposal](https://github.com/zenparsing/es-function-bind), making such functions more readable. For example:
```

Author: Because I'm planning revive `::` notation and plan to present to committee on 2020 Feb meeting, I would like to keep the sentence. What do you think?

Reviewer: You can keep the original, but you should also write about its newer applications. Here are a couple off the top of my head:

- Less transpiling required (TypeScript supports this syntax and it makes sense)
- Now that arrow functions are ubiquitous, a linter could require a named `this` for function declarations.

Author: I'm not sure I understand the second point.

};
```

Explicit `this` parameter also allow type annotation or parameter decorators be added just like normal parameter.

```ts

> // Type annotation (TypeScript)

Reviewer: ```suggestion
// Type annotation (TypeScript, already possible today)
```
